tesla_ver/data_uploading/data_uploading.py

```diff
@@ -0,0 +1,115 @@
+
+        def parse_multiple_contents(contents, filenames):
+            """parses and merges multiple dataframes into a single selection
+
+            Args:
+                contents (list(b64 encoded string)): list of base64 encoded strings representing dataframes from dash upload
+                filenames (list(str)): list of file names corresponding to each dataframe
+            """
+            pass
+
+        df = pd.DataFrame()
+        if None in [contents, filename]:
+            raise PreventUpdate
+        if len(contents) > 1:
+            parse_multiple_contents(contents, filename)
+        else:
+            df = upload_string_to_df(*contents)
+            df[df < 0] = 0
+            df[~df.isin([np.nan, np.inf, -np.inf]).any(1)]
+            df = df.fillna(0)
+            return [
+                {"visibility": "visible"},
+                df.to_dict("records"),
+                [{"name": i, "id": i, "deletable": True, "selectable": True} for i in df.columns],
+            ]
+
+    @app.callback(
+        Output("data-table-visualization", "style_data_conditional"),
+        [Input("data-table-visualization", "selected_columns"), Input("data-table-visualization", "selected_rows"),],
+    )
+    def update_styles(selected_columns, selected_rows):
+        # Generates a list of dictionaries that style the selected rows and columns with a light blue highlight
+        return [
+            *[{"if": {"column_id": i}, "background_color": "#D2F3FF"} for i in selected_columns],
+            *[{"if": {"row_index": i}, "background_color": "#D2F3FF"} for i in selected_rows],
+        ]
+
+    @app.callback(
+        Output("upload-confirmation", "style"),
+        [Input("redis-upload-button", "n_clicks")],
+        [
+            State("data-table-visualization", "data"),
+            State("data-table-visualization", "selected_columns"),
+            State("data-table-visualization", "selected_rows"),
+        ],
+    )
+    def push_to_redis(button_clicks, data_dict, selected_columns, selected_rows):
+        if None in [button_clicks, data_dict, selected_columns]:
+            raise PreventUpdate
+
+        df = (
+            pd.DataFrame.from_records(data_dict)
+            if len(selected_rows) == 0
+            else pd.DataFrame.from_records(data_dict).iloc[selected_rows]
+        )
+
+        mdata_cols = sorted(set(df.columns) - set(selected_columns))
+
+        serialization_context = pa.default_serialization_context()
+
+        redis_manager.redis.set(
+            "data_numeric",
+            serialization_context.serialize(df[sorted(set(["Year", "Subject", *selected_columns]))])
+            .to_buffer()
+            .to_pybytes(),
+        )
+        redis_manager.redis.set(
+            "data_mdata",
+            serialization_context.serialize(df[["Year", "Subject", *mdata_cols]]).to_buffer().to_pybytes(),
+        )
+        redis_manager.redis.set("test", "Hello World, this is a test!")
```

Comment:
What is redis_manager?

Reply:
`redis_manager` is a class that handles connecting to the redis data store once, and then can be imported anywhere to use the redis connection without connecting again every time.  It's located in `data_uploading/redis_manager.py`